monai/data/dataset.py

```diff
@@ -71,6 +72,34 @@ def __getitem__(self, index: int):
         return data
 
 
+class IterableDataset(_TorchIterableDataset):
+    """
+    A generic dataset for iterable data source and an optional callable data transform
+    when fetching a data sample.
+    For example, typical input data can be web data stream which can support multi-process access.
+
+    Note that when used with `DataLoader` and `num_workers > 0`, each worker process will have a
+    different copy of the dataset object, need to guarantee process-safe from data source or DataLoader.
+
+    """
+
+    def __init__(self, data: Sequence, transform: Optional[Callable] = None) -> None:
+        """
+        Args:
+            data: input data source to load and transform to generate dataset for model.
+            transform: a callable data transform on input data.
+        """
+        self.data = iter(data)
+        self.transform = transform
+
+    def __iter__(self):
```

Comment:
```suggestion
    def __iter__(self):
        data_iter = iter(data)
```

the webdataset library stores the iterator as `self.source` as well: https://github.com/tmbdev/webdataset/blob/master/webdataset/dataset.py#L498

maybe it's needed,I don't have a use case at hand

Reply:
updated in the latest commit.
We can adjust more when more use cases come.

Thanks.